src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -0,0 +1,529 @@
+            // we successfully receive or not.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            using (cancellationToken.Register(() => tcsReceive.TrySetCanceled()))
+            {
+                if (_bufferedPayload == null)
+                    _bufferedPayload = await _receiveMessageQueue.DequeuePayloadAsync(cancellationToken).ConfigureAwait(continueOnCapturedContext: true);
+
+                try
+                {
+                    var endOfMessage = _bufferedPayload.BufferPayload(buffer, out WebSocketReceiveResult receiveResult);
+
+                    tcsReceive.SetResult(receiveResult);
+
+                    if (endOfMessage)
+                        _bufferedPayload = null;
+                }
+                catch (Exception exc)
+                {
+                    throw new WebSocketException(WebSocketError.NativeError, exc);
+                }
+
+                return await tcsReceive.Task.ConfigureAwait(continueOnCapturedContext: true);
+            }
+        }
+
+        /// <summary>
+        /// Aborts the connection and cancels any pending IO operations.
+        /// </summary>
+        public override void Abort()
+        {
+            if (_state == _disposed)
+            {
+                return;
+            }
+            _state = (int)WebSocketState.Aborted;
+            Dispose();
+        }
+
+        public override async Task CloseAsync(WebSocketCloseStatus closeStatus, string? statusDescription, CancellationToken cancellationToken)
+        {
+            _writeBuffer = null;
+            ThrowIfNotConnected();
+
+            await CloseAsyncCore(closeStatus, statusDescription, cancellationToken).ConfigureAwait(continueOnCapturedContext: true);
+        }
+
+        private async Task CloseAsyncCore(WebSocketCloseStatus closeStatus, string? statusDescription, CancellationToken cancellationToken)
+        {
+            ThrowOnInvalidState(State,
+                WebSocketState.Open, WebSocketState.CloseReceived, WebSocketState.CloseSent);
+
+            WebSocketValidate.ValidateCloseStatus(closeStatus, statusDescription);
+
+            _tcsClose = new TaskCompletionSource<bool>();
+            // Wrap the cancellationToken in a using so that it can be disposed of whether
+            // we successfully connected or failed trying.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            // In the failure case we need to release the references and dispose of the objects.
+            using (cancellationToken.Register(() => _tcsClose.TrySetCanceled()))
+            {
+                _innerWebSocketCloseStatus = closeStatus;
+                _innerWebSocketCloseStatusDescription = statusDescription;
+                _innerWebSocket!.Invoke("close", (int)closeStatus, statusDescription);
+                await _tcsClose.Task.ConfigureAwait(continueOnCapturedContext: true);
```

Comment:
`return _tcsClose.Task` instead

Reply:
> return _tcsClose.Task instead

Not if the intention is for the cancellation registration to persist until after this completes.

But I have the same concern here that I've voiced elsewhere: this pattern if calling TrySetCanceled in response to cancellation is only causing the caller to wake up, it's not actually canceling the underlying operation, and so the caller will think the operation has completed when it may not have.  In general that can lead to bad race conditions.